src/csharp/Microsoft.Spark/Sql/RowCollector.cs

```diff
@@ -37,5 +38,81 @@ public IEnumerable<Row> Collect(ISocketWrapper socket)
                 }
             }
         }
+
+        /// <summary>
+        /// Synchronously collects pickled row objects from the given socket.
+        /// </summary>
+        /// <param name="socket">Socket the get the stream from</param>
+        /// <param name="server">The jvm socket auth server</param>
+        /// <returns>Collection of row objects</returns>
+        public IEnumerable<Row> SynchronousCollect(ISocketWrapper socket, JvmObjectReference server) =>
+            new SynchronousRowCollector(socket, server);
+
+        /// <summary>
+        /// SynchronousRowCollector synchronously collects Row objects from a socket.
+        /// </summary>
+        private class SynchronousRowCollector : IEnumerable<Row>
+        {
+            private readonly ISocketWrapper _socket;
+            private readonly JvmObjectReference _server;
+
+            private int _readStatus = 1;
+            private IEnumerable<Row> _collectEnumerable = null;
+
+            internal SynchronousRowCollector(ISocketWrapper socket, JvmObjectReference server)
+            {
+                _socket = socket;
+                _server = server;
+            }
+
+            ~SynchronousRowCollector()
+            {
+                // If iterator is not fully consumed
+                if ((_readStatus == 1) && (_collectEnumerable != null))
+                {
+                    // Finish consuming partition data stream
+                    foreach (Row _ in _collectEnumerable)
+                    {
+                    }
+
+                    // Tell Java to stop sending data and close connection
+                    Stream outputStream = _socket.OutputStream;
+                    SerDe.Write(outputStream, 0);
+                    outputStream.Flush();
+                }
+            }
+
+            public IEnumerator<Row> GetEnumerator()
+            {
+                Stream inputStream = _socket.InputStream;
+                Stream outputStream = _socket.OutputStream;
+
+                while (_readStatus == 1)
+                {
+                    // Request next partition data from Java
+                    SerDe.Write(outputStream, 1);
+                    outputStream.Flush();
+
+                    // If response is 1 then there is a partition to read, if 0 then fully consumed
```

Comment:
then should we add `else` and assert the `_readStatus` is `0` to detect any issues early?